lib/fluent/plugin/out_webhdfs.rb

```diff
@@ -247,13 +264,52 @@ def send_data(path, data)
     end
   end
 
+  HOSTNAME_PLACEHOLDERS_DEPRECATED = ['${hostname}', '%{hostname}', '__HOSTNAME__']
+  UUID_RANDOM_PLACEHOLDERS_DEPRECATED = ['${uuid}', '${uuid:random}', '__UUID__', '__UUID_RANDOM__']
+  UUID_OTHER_PLACEHOLDERS_OBSOLETED = ['${uuid:hostname}', '%{uuid:hostname}', '__UUID_HOSTNAME__', '${uuid:timestamp}', '%{uuid:timestamp}', '__UUID_TIMESTAMP__']
+
+  def verify_config_placeholders_in_path!(conf)
+    return unless conf.has_key?('path')
+
+    path = conf['path']
+
+    # check @path for ${hostname}, %{hostname} and __HOSTNAME__ to warn to use #{Socket.gethostbyname}
+    if HOSTNAME_PLACEHOLDERS_DEPRECATED.any?{|ph| path.include?(ph) }
+      log.warn "hostname placeholder is now deprecated. use '\#\{Socket.gethostname\}' instead."
+      hostname = conf['hostname'] || Socket.gethostname
+      HOSTNAME_PLACEHOLDERS_DEPRECATED.each do |ph|
+        path.gsub!(ph, hostname)
+      end
+    end
+
+    if UUID_RANDOM_PLACEHOLDERS_DEPRECATED.any?{|ph| path.include?(ph) }
+      log.warn "random uuid placeholders are now deprecated. use %{uuid} (or %{uuid_flush}) instead."
+      UUID_RANDOM_PLACEHOLDERS_DEPRECATED.each do |ph|
+        path.gsub!(ph, '%{uuid}')
+      end
+    end
+
+    if UUID_OTHER_PLACEHOLDERS_OBSOLETED.any?{|ph| path.include?(ph) }
+      UUID_OTHER_PLACEHOLDERS_OBSOLETED.each do |ph|
+        if path.include?(ph)
+          log.error "configuration placeholder #{ph} is now unsupported by webhdfs output plugin."
+        end
+      end
+      raise ConfigError, "there are unsupported placeholders in path."
+    end
+  end
+
   def generate_path(chunk)
     hdfs_path = if @append
                   extract_placeholders(@path, chunk.metadata)
                 else
                   extract_placeholders(@path, chunk.metadata).gsub(CHUNK_ID_PLACE_HOLDER, dump_unique_id(chunk.unique_id))
                 end
     hdfs_path = "#{hdfs_path}#{@compressor.ext}"
+    if @replace_random_uuid
+      uuid_random = SecureRandom.uuid
+      hdfs_path.gsub!('%{uuid}', uuid_random).gsub!('%{uuid_flush}', uuid_random)
```

Comment:
Is `gsub!` chaining safe?
`gsub!` returns `nil` when substitution doesn't happen.